README.md
# sonarqube_exporter

An exporter of metrics for each project in Sonarqube.

## Usage

```
./sonarqube_exporter -h
Usage of ./sonarqube_exporter:
  -http.address string
    	Address to listen on (default ":9344")
  -http.telemetry-path string
    	Path under which the exporter exposes its metrics (default "/metrics")
  -log.level string
    	Log level (default "ERROR")
  -sonarqube.password string
    	Password to use for authentication
  -sonarqube.project-filter string
    	Regexp to limit the number of projects to scrape. Applied to the key of each project. (default ".*")
  -sonarqube.url string
    	URL of Sonarqube (default "http://localhost:8080")
  -sonarqube.username string
    	Username to use for authentication
```

## Development

### Building

```
$ go build
```

### Sonarqube API documentation

API docs can be found in the local installation of Sonarqube at `http://<sonarqube>/web_api`.


CHANGELOG.md
## 0.3.0

* [CHANGE] Support metrics of type RATING and LEVEL

## 0.2.0

* [FEATURE] Export metrics `alert_status`, `quality_gate_details` and `reliability_rating`
* [CHANGE] Use `promhttp.Handler` instead of deprecated `prometheus.Handler`

## 0.1.0

Initial release



main_test.go
package main

import (
	"bytes"
	"context"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"os/exec"
	"testing"
	"time"
)

func TestExporter(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/api/metrics/search":
			page := r.URL.Query().Get("p")
			b, err := ioutil.ReadFile(fmt.Sprintf("fixtures/metrics_search_%s.json", page))
			if err != nil {
				w.WriteHeader(http.StatusInternalServerError)
				w.Write([]byte(err.Error()))
				return
			}
			w.WriteHeader(http.StatusOK)
			w.Write(b)
		case "/api/components/search":
			page := r.URL.Query().Get("p")
			b, err := ioutil.ReadFile(fmt.Sprintf("fixtures/components_search_%s.json", page))
			if err != nil {
				w.WriteHeader(http.StatusInternalServerError)
				w.Write([]byte(err.Error()))
				return
			}
			w.WriteHeader(http.StatusOK)
			w.Write(b)
		case "/api/measures/component":
			cID := r.URL.Query().Get("componentId")
			b, err := ioutil.ReadFile(fmt.Sprintf("fixtures/measures_component_%s.json", cID))
			if err != nil {
				w.WriteHeader(http.StatusInternalServerError)
				w.Write([]byte(err.Error()))
				return
			}
			w.WriteHeader(http.StatusOK)
			w.Write(b)
		}
	}))
	defer ts.Close()

	ctx, cancel := context.WithCancel(context.Background())
	cmd := exec.CommandContext(ctx, "./sonarqube_exporter", "-sonarqube.url", ts.URL, "-log.level", "debug")
	var stderr, stdout bytes.Buffer
	cmd.Stderr = &stderr
	cmd.Stdout = &stdout
	defer cancel()
	go func() {
		err := cmd.Run()
		if err != nil {
			t.Fatal(err)
		}
	}()

	time.Sleep(1 * time.Second)
	resp, err := http.Get("http://localhost:9344/metrics")
	if err != nil {
		t.Fatal(err)
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		t.Fatal(err)
	}

	tests := []string{
		`sonarqube_exporter_scrapes_total 1`,
		`sonarqube_up 1`,
		`sonarqube_measures{component_key="identifier:testOne",metric="branch_coverage"} 35.4`,
		`sonarqube_measures{component_key="identifier:testOne",metric="conditions_to_cover"} 794`,
		`sonarqube_measures{component_key="identifier:testTwo",metric="class_complexity"} 4.9`,
		`sonarqube_measures{component_key="identifier:testOne",metric="alert_status"} 2`,
		`sonarqube_measures{component_key="identifier:testTwo",metric="reliability_rating"} 1`,
		`sonarqube_measures{component_key="identifier:testTwo",metric="quality_gate_details"} 0`,
		`sonarqube_measures{component_key="identifier:testOne",metric="reliability_rating"} 0`,
	}
	for _, test := range tests {
		if !bytes.Contains(body, []byte(test)) {
			t.Errorf("want metrics to include %q, have:\n%s", test, body)
			t.Errorf("STDOUT:\n%s", string(stdout.Bytes()))
			t.Errorf("STDERR:\n%s", string(stderr.Bytes()))
		}
	}
}


sonarqube.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"strings"
	"time"

	log "github.com/sirupsen/logrus"
)

type metricsSearchResponse struct {
	Metrics  []*metric `json:"metrics"`
	Total    int       `json:"total"`
	Page     int       `json:"p"`
	PageSize int       `json:"ps"`
}

type metric struct {
	Key  string `json:"key"`
	Type string `json:"type"`
}

type componentMeasuresResponse struct {
	Component *componentMeasureItem
}

type componentMeasureItem struct {
	ID       string
	Key      string
	Measures []*measure
}

type componentSearchResponse struct {
	Paging     *componentSearchPaging
	Components []*componentSearchItem
}

type componentSearchPaging struct {
	PageIndex int
	PageSize  int
	Total     int
}

type componentSearchItem struct {
	ID  string `json:"id"`
	Key string `json:"key"`
}

type measure struct {
	Metric string
	Value  string
}

type apiClient struct {
	client       *http.Client
	password     string
	sonarqubeURL string
	username     string
}

func (a *apiClient) findAllMetrics() ([]*metric, error) {
	page := 1
	pageSize := 100
	metrics := []*metric{}
	var total int
	for {
		msr := &metricsSearchResponse{}
		url := fmt.Sprintf("%s/api/metrics/search?ps=%d&p=%d", a.sonarqubeURL, pageSize, page)
		err := a.requestJSON(url, msr)
		if err != nil {
			return nil, err
		}

		pageSize = msr.PageSize
		total = msr.Total
		metrics = append(metrics, msr.Metrics...)

		if (page * pageSize) >= total {
			return metrics, nil
		}

		page = page + 1
	}
}

func (a *apiClient) findAllProjects() ([]*componentSearchItem, error) {
	pageIndex := 1
	pageSize := 100
	components := []*componentSearchItem{}
	for {
		csr := &componentSearchResponse{}
		url := fmt.Sprintf("%s/api/components/search?ps=%d&p=%d&qualifiers=TRK", a.sonarqubeURL, pageSize, pageIndex)
		err := a.requestJSON(url, csr)
		if err != nil {
			return nil, err
		}

		pageSize = csr.Paging.PageSize
		components = append(components, csr.Components...)
		if (pageIndex * pageSize) >= csr.Paging.Total {
			return components, nil
		}

		pageIndex = csr.Paging.PageIndex + 1
	}
}

func (a *apiClient) findMeasuresForComponent(componentID string, metricKeys []string) (*componentMeasuresResponse, error) {
	url := fmt.Sprintf("%s/api/measures/component?componentId=%s&metricKeys=%s", a.sonarqubeURL, componentID, strings.Join(metricKeys, ","))
	cmr := &componentMeasuresResponse{}
	err := a.requestJSON(url, cmr)
	if err != nil {
		return nil, err
	}

	return cmr, nil
}

func (a *apiClient) requestJSON(url string, v interface{}) error {
	log.Debugf("Sending request to '%s'", url)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return err
	}

	if a.password != "" && a.username != "" {
		log.Debug("Setting basic auth")
		req.SetBasicAuth(a.username, a.password)
	}

	resp, err := a.client.Do(req)
	if err != nil {
		return err
	}

	defer resp.Body.Close()
	data, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	err = json.Unmarshal(data, v)
	if err != nil {
		return err
	}

	return nil
}

func newAPIClient(h *http.Client, username string, password string, url string) *apiClient {
	if h == nil {
		h = &http.Client{Timeout: 1 * time.Second}
	}

	return &apiClient{
		client:       h,
		password:     password,
		sonarqubeURL: url,
		username:     username,
	}
}


main.go
package main

import (
	"flag"
	"net/http"
	"regexp"
	"strconv"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	log "github.com/sirupsen/logrus"
)

const (
	namespace = "sonarqube"
)

var (
	httpAddress         = flag.String("http.address", ":9344", "Address to listen on")
	httpTelemetryPath   = flag.String("http.telemetry-path", "/metrics", "Path under which the exporter exposes its metrics")
	logLevel            = flag.String("log.level", "ERROR", "Log level")
	sonarqubePassword   = flag.String("sonarqube.password", "", "Password to use for authentication")
	sonarqubeURL        = flag.String("sonarqube.url", "http://localhost:8080", "URL of Sonarqube")
	sonarqubeUsername   = flag.String("sonarqube.username", "", "Username to use for authentication")
	projectFilterRegex  = flag.String("sonarqube.project-filter", ".*", "Regexp to limit the number of projects to scrape. Applied to the key of each project.")
	accpetedMetricTypes = map[string]struct{}{"INT": {}, "PERCENT": {}, "FLOAT": {}, "DATA": {}, "RATING": {}, "LEVEL": {}}
)

type exporter struct {
	client         *apiClient
	projectFilter  *regexp.Regexp
	projectMetrics *prometheus.GaugeVec
	up             prometheus.Gauge
	totalScrapes   prometheus.Counter
}

func (e *exporter) Collect(ch chan<- prometheus.Metric) {
	e.scrape()
	e.projectMetrics.Collect(ch)
	ch <- e.totalScrapes
	ch <- e.up
}

func (e *exporter) Describe(ch chan<- *prometheus.Desc) {
	e.projectMetrics.Describe(ch)
	ch <- e.totalScrapes.Desc()
	ch <- e.up.Desc()
}

func (e *exporter) scrape() {
	e.totalScrapes.Inc()
	allMetrics, err := e.client.findAllMetrics()
	if err != nil {
		log.Errorf("Finding all metrics: %s ", err)
		e.up.Set(0)
		return
	}

	log.Debugf("Found %d metrics", len(allMetrics))
	metricKeys := []string{}
	for _, m := range allMetrics {
		if _, exists := accpetedMetricTypes[m.Type]; exists {
			metricKeys = append(metricKeys, m.Key)
		}
	}

	allProjects, err := e.client.findAllProjects()
	if err != nil {
		log.Errorf("Finding all projects: %s ", err)
		e.up.Set(0)
		return
	}

	log.Debugf("Found %d projects", len(allProjects))
	for _, p := range allProjects {
		if !e.projectFilter.MatchString(p.Key) {
			continue
		}

		log.Debugf("Requesting metrics for %s...", p.ID)
		r, err := e.client.findMeasuresForComponent(p.ID, metricKeys)
		if err != nil {
			log.Errorf("Finding measures for component '%s': %s", p.ID, err)
			e.up.Set(0)
			return
		}

		for _, measure := range r.Component.Measures {
			var measureFloat float64
			if metric, exists := dataMetricsValues[measure.Metric]; exists {
				var valueExists bool
				measureFloat, valueExists = metric[measure.Value]
				if !valueExists {
					continue
				}
			} else if measure.Value != "" {
				if measureFloat, err = strconv.ParseFloat(measure.Value, 64); err != nil {
					log.Debugf("Value of measure '%s' could not be parsed: %s", measure.Metric, err)
					continue
				}
			}

			e.projectMetrics.WithLabelValues(p.Key, measure.Metric).Set(measureFloat)
		}
	}

	e.up.Set(1)
}

func mustInitLogging(lvl string) {
	level, err := log.ParseLevel(lvl)
	if err != nil {
		log.Fatal(err)
	}

	log.SetLevel(level)
}

func main() {
	flag.Parse()
	mustInitLogging(*logLevel)
	e := &exporter{
		client: newAPIClient(nil, *sonarqubeUsername, *sonarqubePassword, *sonarqubeURL),
		up: prometheus.NewGauge(prometheus.GaugeOpts{
			Namespace: namespace,
			Name:      "up",
			Help:      "Was the last scrape of Sonarqube successful.",
		}),
		projectFilter: regexp.MustCompile(*projectFilterRegex),
		projectMetrics: prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Namespace:   namespace,
				Name:        "measures",
				Help:        "A measure of a project in Sonarqube.",
				ConstLabels: nil,
			},
			[]string{"component_key", "metric"},
		),
		totalScrapes: prometheus.NewCounter(prometheus.CounterOpts{
			Namespace: namespace,
			Name:      "exporter_scrapes_total",
			Help:      "Total scrapes of the Sonarqube exporter.",
		}),
	}
	prometheus.MustRegister(e)
	http.Handle(*httpTelemetryPath, promhttp.Handler())
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`<html>
<head><title>Sonarqube Exporter</title></head>
<body>
<h1>Sonarqube Exporter</h1>
<p><a href='` + *httpTelemetryPath + `'>Metrics</a></p>
</body>
</html>`))
	})
	log.Fatal(http.ListenAndServe(*httpAddress, nil))
}
